eessi/reframe/eessi_utils/hooks.py

```diff
@@ -15,109 +17,165 @@
+    - num_cpus_per_task = cpus_per_node / num_tasks_per_node
 
-    If num_tasks_per_node is set, set num_gpus_per_node equal to either num_tasks_per_node,
-    or nb of GPUs per node available in this partition (whatever is smallest).
+    If num_tasks_per_node is set, set num_gpus_per_node equal to either num_tasks_per_node or gpus_per_node
+    (whichever is smallest), unless num_gpus_per_node is also set.
     """
-    max_gpus_per_node = utils.get_num_gpus_per_node(test)
-    max_cpus_per_node = test.max_cpus_per_node
-    num_tasks_per_node = test.num_tasks_per_node
-    num_gpus_per_node = test.num_gpus_per_node
+    max_avail_gpus_per_node = utils.get_max_avail_gpus_per_node(test)
+
+    # Check if the default number of gpus per node is already defined in the test
+    # (e.g. by earlier hooks like set_tag_scale).
+    # If so, check if it doesn't exceed the maximum available.
+    # If not, set default_num_gpus_per_node based on the maximum available gpus and node_part
+    if test.default_num_gpus_per_node:
+        # may skip if not enough GPUs
+        test.skip_if(
+            test.default_num_gpus_per_node > max_avail_gpus_per_node,
+            f'GPUs per node ({test.default_num_gpus_per_node}) is higher than max available ({max_avail_gpus_per_node})'
+            f' in current partition ({test.current_partition.name})'
+        )
+    else:
+        # no default set yet, so setting one
+        test.default_num_gpus_per_node = math.ceil(max_avail_gpus_per_node / test.node_part)
 
-    if not num_tasks_per_node:
-        if not num_gpus_per_node:
-            num_gpus_per_node = max_gpus_per_node
-        num_tasks_per_node = num_gpus_per_node
+    # neither num_tasks_per_node nor num_gpus_per_node are set
+    if not test.num_tasks_per_node and not test.num_gpus_per_node:
+        test.num_gpus_per_node = test.default_num_gpus_per_node
+        test.num_tasks_per_node = test.num_gpus_per_node
 
-    elif not num_gpus_per_node:
-        num_gpus_per_node = min(num_tasks_per_node, max_gpus_per_node)
+    # num_tasks_per_node is not set, but num_gpus_per_node is
+    elif not test.num_tasks_per_node:
+        test.num_tasks_per_node = test.num_gpus_per_node
 
+    # num_gpus_per_node is not set, but num_tasks_per_node is
+    elif not test.num_gpus_per_node:
+        test.num_gpus_per_node = min(test.num_tasks_per_node, test.default_num_gpus_per_node)
+
+    else:
+        pass  # both num_tasks_per_node and num_gpus_per_node are already set
+
+    # num_cpus_per_task is not set
     if not test.num_cpus_per_task:
-        test.num_cpus_per_task = int(
-            (max_cpus_per_node * num_gpus_per_node) / (num_tasks_per_node * max_gpus_per_node)
+        max_avail_cpus_per_gpu = int(test.max_avail_cpus_per_node / max_avail_gpus_per_node)
+        test.num_cpus_per_task = min(
+            int(test.default_num_cpus_per_node / test.num_tasks_per_node),
+            max_avail_cpus_per_gpu
         )
```

Comment:
num_cpus_per_task was not set correctly in all cases, so i had to change the logic here.
i'm pretty sure it's fixed now, but would be good to double check.